core/domain/question_services.py

```diff
@@ -201,3 +203,39 @@ def update_question(committer_id, question_id, change_list, commit_message):
     """
     question = apply_change_list(question_id, change_list)
     _save_question(committer_id, question, change_list, commit_message)
+
+
+def get_questions_batch(
+        collection_id, skill_ids, user_id, question_play_counts, batch_size):
+    """Fetches a batch of questions.
+
+    Args:
+        collection_id: str. Id of the collection.
+        skill_ids: list(str). A list of skill ids.
+        user_id: str. Id of the user.
+        question_play_counts: dict. A dict mapping question ids to the number
+            of times it has been played in this question session.
+        batch_size: int. The intende number of questions to be returned.
+
+    Returns:
+        A list of Question objects.
+    """
+    user_skill_ids = collection_services.get_acquired_skills_of_user(user_id)
+    question_skill_ids = list(set(user_skill_ids) & set(skill_ids))
+
+    collection = collection_services.get_collection_by_id(collection_id)
+    question_ids = []
+    for skill_id in question_skill_ids:
+        if collection.skills[skill_id]:
+            question_ids.extend(collection.skills[skill_id].question_ids)
+    unique_question_ids = list(set(question_ids))
+
+    if len(unique_question_ids) < batch_size:
+        batch_size = len(unique_question_ids)
+    random_question_ids = random.sample(unique_question_ids, batch_size)
+
+    questions_batch = []
+    for question_id in random_question_ids:
+        questions_batch.append(get_question_by_id(question_id))
+        question_play_counts[question_id] += 1
```

Comment:
I don't think this function should touch question_play_counts. Let the caller handle that. It's not really relevant to the purpose of this function, which is simply to get a batch of questions based on the input args.

Reply:
Understood, by the comment below, i am assuming it will be handled in the frontend.

Reply:
Yeah, I think so.